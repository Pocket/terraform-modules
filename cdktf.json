{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
