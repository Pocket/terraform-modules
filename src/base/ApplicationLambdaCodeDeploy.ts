import { CodedeployApp } from '@cdktf/provider-aws/lib/codedeploy-app';
import { CodedeployDeploymentGroup } from '@cdktf/provider-aws/lib/codedeploy-deployment-group';
import { CodestarnotificationsNotificationRule } from '@cdktf/provider-aws/lib/codestarnotifications-notification-rule';
import { DataAwsIamPolicyDocument } from '@cdktf/provider-aws/lib/data-aws-iam-policy-document';
import { IamRole } from '@cdktf/provider-aws/lib/iam-role';
import { IamRolePolicyAttachment } from '@cdktf/provider-aws/lib/iam-role-policy-attachment';
import { TerraformMetaArguments } from 'cdktf';
import { Construct } from 'constructs';

export interface ApplicationVersionedLambdaCodeDeployProps
  extends TerraformMetaArguments {
  name: string;
  deploySnsTopicArn?: string;
  detailType?: 'BASIC' | 'FULL';
  region: string;
  accountId: string;
  tags?: { [key: string]: string };
  notifications?: {
    /**
     * Option to send CodeDeploy notifications on Started event, defaults to false.
     */
    notifyOnStarted?: boolean;
    /**
     * Option to send CodeDeploy notifications on Succeeded event, defaults to false.
     */
    notifyOnSucceeded?: boolean;
    /**
     * Option to send CodeDeploy notifications on Failed event, defaults to true.
     */
    notifyOnFailed?: boolean;
  };
}

export class ApplicationLambdaCodeDeploy extends Construct {
  public readonly codeDeployApp: CodedeployApp;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationVersionedLambdaCodeDeployProps
  ) {
    super(scope, name);

    this.codeDeployApp = this.setupCodeDeploy();
  }

  private setupCodeDeploy() {
    const codeDeployApp = new CodedeployApp(this, 'code-deploy-app', {
      name: `${this.config.name}-Lambda`,
      computePlatform: 'Lambda',
      provider: this.config.provider,
      tags: this.config.tags,
    });

    this.createCodeDeploymentGroup(codeDeployApp);

    if (this.config.deploySnsTopicArn) {
      this.setupCodeDeployNotifications(codeDeployApp);
    }

    return codeDeployApp;
  }

  private createCodeDeploymentGroup(codeDeployApp: CodedeployApp) {
    new CodedeployDeploymentGroup(this, 'code-deployment-group', {
      appName: codeDeployApp.name,
      deploymentConfigName: 'CodeDeployDefault.LambdaAllAtOnce',
      deploymentGroupName: codeDeployApp.name,
      serviceRoleArn: this.getCodeDeployRole().arn,
      deploymentStyle: {
        deploymentType: 'BLUE_GREEN',
        deploymentOption: 'WITH_TRAFFIC_CONTROL',
      },
      autoRollbackConfiguration: {
        enabled: true,
        events: ['DEPLOYMENT_FAILURE'],
      },
      dependsOn: [codeDeployApp],
      provider: this.config.provider,
      tags: this.config.tags,
    });
  }

  private getCodeDeployRole() {
    const codeDeployRole = new IamRole(this, 'code-deploy-role', {
      name: `${this.config.name}-CodeDeployRole`,
      assumeRolePolicy: this.getCodeDeployAssumePolicyDocument(),
      provider: this.config.provider,
      tags: this.config.tags,
    });

    new IamRolePolicyAttachment(this, 'code-deploy-policy-attachment', {
      policyArn:
        'arn:aws:iam::aws:policy/service-role/AWSCodeDeployRoleForLambda',
      role: codeDeployRole.name,
      dependsOn: [codeDeployRole],
      provider: this.config.provider,
    });

    return codeDeployRole;
  }

  private getCodeDeployAssumePolicyDocument() {
    return new DataAwsIamPolicyDocument(
      this,
      'code-deploy-assume-role-policy-document',
      {
        statement: [
          {
            effect: 'Allow',
            actions: ['sts:AssumeRole'],
            principals: [
              {
                identifiers: ['codedeploy.amazonaws.com'],
                type: 'Service',
              },
            ],
          },
        ],
        provider: this.config.provider,
      }
    ).json;
  }

  private setupCodeDeployNotifications(codeDeployApp: CodedeployApp) {
    const eventTypeIds = [];

    // the OR condition here sets the notification for failed deploys which is a default when no notification preferences are provided
    if (
      this.config.notifications?.notifyOnFailed ||
      this.config.notifications?.notifyOnFailed === undefined
    ) {
      eventTypeIds.push('codedeploy-application-deployment-failed');
    }

    if (this.config.notifications?.notifyOnSucceeded) {
      eventTypeIds.push('codedeploy-application-deployment-succeeded');
    }

    if (this.config.notifications?.notifyOnStarted) {
      eventTypeIds.push('codedeploy-application-deployment-started');
    }

    new CodestarnotificationsNotificationRule(this, 'notifications', {
      detailType: this.config.detailType ?? 'BASIC',
      eventTypeIds: eventTypeIds,
      name: codeDeployApp.name,
      resource: `arn:aws:codedeploy:${this.config.region}:${this.config.accountId}:application:${codeDeployApp.name}`,
      target: [
        {
          address: this.config.deploySnsTopicArn,
        },
      ],
      provider: this.config.provider,
      tags: this.config.tags,
    });
  }
}
