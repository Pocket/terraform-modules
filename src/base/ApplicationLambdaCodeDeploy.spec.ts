import { Testing, TerraformStack } from 'cdktf';
import { ApplicationLambdaCodeDeploy } from './ApplicationLambdaCodeDeploy';

const config = {
  name: 'Test-Lambda-Code-Deploy',
  deploySnsTopicName: 'Test-Deployment-Topic',
};

test('renders a lambda code deploy app', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationLambdaCodeDeploy(stack, 'test-lambda-code-deploy', config);

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('renders a lambda code deploy app with detail type', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationLambdaCodeDeploy(stack, 'test-lambda-code-deploy', {
    ...config,
    detailType: 'FULL',
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});
