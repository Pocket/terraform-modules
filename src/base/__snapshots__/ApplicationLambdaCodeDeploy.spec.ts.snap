// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a lambda code deploy app 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_codedeploy_app\\": {
      \\"test_testlambdacodedeploy_codedeployapp_8F9AF4BD\\": {
        \\"compute_platform\\": \\"Lambda\\",
        \\"name\\": \\"Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-app\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployapp_8F9AF4BD\\"
          }
        }
      }
    },
    \\"aws_iam_role\\": {
      \\"test_testlambdacodedeploy_codedeployrole_755F51E7\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46.json}\\",
        \\"name\\": \\"Test-Lambda-Code-Deploy-CodeDeployRole\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-role\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployrole_755F51E7\\"
          }
        }
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test_testlambdacodedeploy_codedeploypolicyattachment_09FCA989\\": {
        \\"policy_arn\\": \\"arn:aws:iam::aws:policy/service-role/AWSCodeDeployRoleForLambda\\",
        \\"role\\": \\"Test-Lambda-Code-Deploy-CodeDeployRole\\",
        \\"depends_on\\": [
          \\"aws_iam_role.test_testlambdacodedeploy_codedeployrole_755F51E7\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-policy-attachment\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeploypolicyattachment_09FCA989\\"
          }
        }
      }
    },
    \\"aws_codedeploy_deployment_group\\": {
      \\"test_testlambdacodedeploy_codedeploymentgroup_A7E2E869\\": {
        \\"app_name\\": \\"Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"deployment_config_name\\": \\"CodeDeployDefault.LambdaAllAtOnce\\",
        \\"deployment_group_name\\": \\"Test-Lambda-Code-Deploy\\",
        \\"service_role_arn\\": \\"\${aws_iam_role.test_testlambdacodedeploy_codedeployrole_755F51E7.arn}\\",
        \\"auto_rollback_configuration\\": [
          {
            \\"enabled\\": true,
            \\"events\\": [
              \\"DEPLOYMENT_FAILURE\\"
            ]
          }
        ],
        \\"deployment_style\\": [
          {
            \\"deployment_type\\": \\"BLUE_GREEN\\",
            \\"deployment_option\\": \\"WITH_TRAFFIC_CONTROL\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deployment-group\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeploymentgroup_A7E2E869\\"
          }
        }
      }
    }
  },
  \\"data\\": {
    \\"aws_iam_policy_document\\": {
      \\"test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46\\": {
        \\"statement\\": [
          {
            \\"effect\\": \\"Allow\\",
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"codedeploy.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-assume-role-policy-document\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46\\"
          }
        }
      }
    }
  }
}"
`;

exports[`renders a lambda code deploy app with sns topic arn 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_codedeploy_app\\": {
      \\"test_testlambdacodedeploy_codedeployapp_8F9AF4BD\\": {
        \\"compute_platform\\": \\"Lambda\\",
        \\"name\\": \\"Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-app\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployapp_8F9AF4BD\\"
          }
        }
      }
    },
    \\"aws_iam_role\\": {
      \\"test_testlambdacodedeploy_codedeployrole_755F51E7\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46.json}\\",
        \\"name\\": \\"Test-Lambda-Code-Deploy-CodeDeployRole\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-role\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployrole_755F51E7\\"
          }
        }
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test_testlambdacodedeploy_codedeploypolicyattachment_09FCA989\\": {
        \\"policy_arn\\": \\"arn:aws:iam::aws:policy/service-role/AWSCodeDeployRoleForLambda\\",
        \\"role\\": \\"Test-Lambda-Code-Deploy-CodeDeployRole\\",
        \\"depends_on\\": [
          \\"aws_iam_role.test_testlambdacodedeploy_codedeployrole_755F51E7\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-policy-attachment\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeploypolicyattachment_09FCA989\\"
          }
        }
      }
    },
    \\"aws_codedeploy_deployment_group\\": {
      \\"test_testlambdacodedeploy_codedeploymentgroup_A7E2E869\\": {
        \\"app_name\\": \\"Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"deployment_config_name\\": \\"CodeDeployDefault.LambdaAllAtOnce\\",
        \\"deployment_group_name\\": \\"Test-Lambda-Code-Deploy\\",
        \\"service_role_arn\\": \\"\${aws_iam_role.test_testlambdacodedeploy_codedeployrole_755F51E7.arn}\\",
        \\"auto_rollback_configuration\\": [
          {
            \\"enabled\\": true,
            \\"events\\": [
              \\"DEPLOYMENT_FAILURE\\"
            ]
          }
        ],
        \\"deployment_style\\": [
          {
            \\"deployment_type\\": \\"BLUE_GREEN\\",
            \\"deployment_option\\": \\"WITH_TRAFFIC_CONTROL\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deployment-group\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeploymentgroup_A7E2E869\\"
          }
        }
      }
    },
    \\"aws_codestarnotifications_notification_rule\\": {
      \\"test_testlambdacodedeploy_notifications_81C5AFD8\\": {
        \\"detail_type\\": \\"BASIC\\",
        \\"event_type_ids\\": [
          \\"codedeploy-application-deployment-failed\\"
        ],
        \\"name\\": \\"Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"resource\\": \\"arn:aws:codedeploy:us-east-1:123:application:Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"target\\": [
          {
            \\"address\\": \\"test:deploy-topic:arn\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/notifications\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_notifications_81C5AFD8\\"
          }
        }
      }
    }
  },
  \\"data\\": {
    \\"aws_iam_policy_document\\": {
      \\"test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46\\": {
        \\"statement\\": [
          {
            \\"effect\\": \\"Allow\\",
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"codedeploy.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-assume-role-policy-document\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46\\"
          }
        }
      }
    }
  }
}"
`;

exports[`renders a lambda code deploy app with sns topic arn and detail type 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_codedeploy_app\\": {
      \\"test_testlambdacodedeploy_codedeployapp_8F9AF4BD\\": {
        \\"compute_platform\\": \\"Lambda\\",
        \\"name\\": \\"Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-app\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployapp_8F9AF4BD\\"
          }
        }
      }
    },
    \\"aws_iam_role\\": {
      \\"test_testlambdacodedeploy_codedeployrole_755F51E7\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46.json}\\",
        \\"name\\": \\"Test-Lambda-Code-Deploy-CodeDeployRole\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-role\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployrole_755F51E7\\"
          }
        }
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test_testlambdacodedeploy_codedeploypolicyattachment_09FCA989\\": {
        \\"policy_arn\\": \\"arn:aws:iam::aws:policy/service-role/AWSCodeDeployRoleForLambda\\",
        \\"role\\": \\"Test-Lambda-Code-Deploy-CodeDeployRole\\",
        \\"depends_on\\": [
          \\"aws_iam_role.test_testlambdacodedeploy_codedeployrole_755F51E7\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-policy-attachment\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeploypolicyattachment_09FCA989\\"
          }
        }
      }
    },
    \\"aws_codedeploy_deployment_group\\": {
      \\"test_testlambdacodedeploy_codedeploymentgroup_A7E2E869\\": {
        \\"app_name\\": \\"Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"deployment_config_name\\": \\"CodeDeployDefault.LambdaAllAtOnce\\",
        \\"deployment_group_name\\": \\"Test-Lambda-Code-Deploy\\",
        \\"service_role_arn\\": \\"\${aws_iam_role.test_testlambdacodedeploy_codedeployrole_755F51E7.arn}\\",
        \\"auto_rollback_configuration\\": [
          {
            \\"enabled\\": true,
            \\"events\\": [
              \\"DEPLOYMENT_FAILURE\\"
            ]
          }
        ],
        \\"deployment_style\\": [
          {
            \\"deployment_type\\": \\"BLUE_GREEN\\",
            \\"deployment_option\\": \\"WITH_TRAFFIC_CONTROL\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deployment-group\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeploymentgroup_A7E2E869\\"
          }
        }
      }
    },
    \\"aws_codestarnotifications_notification_rule\\": {
      \\"test_testlambdacodedeploy_notifications_81C5AFD8\\": {
        \\"detail_type\\": \\"FULL\\",
        \\"event_type_ids\\": [
          \\"codedeploy-application-deployment-failed\\"
        ],
        \\"name\\": \\"Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"resource\\": \\"arn:aws:codedeploy:us-east-1:123:application:Test-Lambda-Code-Deploy-CodeDeploy\\",
        \\"target\\": [
          {
            \\"address\\": \\"test:deploy-topic:arn\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/notifications\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_notifications_81C5AFD8\\"
          }
        }
      }
    }
  },
  \\"data\\": {
    \\"aws_iam_policy_document\\": {
      \\"test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46\\": {
        \\"statement\\": [
          {
            \\"effect\\": \\"Allow\\",
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"codedeploy.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-lambda-code-deploy/code-deploy-assume-role-policy-document\\",
            \\"uniqueId\\": \\"test_testlambdacodedeploy_codedeployassumerolepolicydocument_95061B46\\"
          }
        }
      }
    }
  }
}"
`;
